core/app/models/spree/order_shipping.rb

```diff
@@ -71,17 +71,12 @@ def ship(inventory_units:, stock_location:, address:, shipping_method:,
       shipment.update_columns(state: 'shipped', shipped_at: Time.current)
     end
 
-    send_shipment_emails(carton) if stock_location.fulfillable? && !suppress_mailer # e.g. digital gift cards that aren't actually shipped
+    Spree.event_bus.publish(
+      :carton_shipped,
+      Spree::Events::CartonShippedEvent.new(carton: carton, suppress_mailer: suppress_mailer)
```

Comment:
I don't like our event depend on `suppress_mailer` here. Can't we just pass the `order` instead?

Reply:
Or actually pass the `carton` - it holds a reference to the order(s).

Reply:
It is already passing the `carton` and as the carton only nows it's `orders` we do not know the `order` object that sends this event, though.

Reply:
I agree this is not ideal. I've removed `suppress_mailer` from the event object, and have made a (possibly presumptuous) decision within the `MailProcessor`:
```
return if event.carton.shipments.any?(&:suppress_mailer)
```
The `suppress_mailer` flag is defined on `Shipment`. Having the carton and order in the event does not give the necessary context of which _shipment_ just triggered this `CartonShippedEvent`.

Thoughts?

Reply:
Do not send the event if suppress_mailer is true? This is it’s whole purpose right? “Do not send” as in “do not notify the user”, whatever the method of notification is.

Reply:
That works for customer-related notifications, but there could possibly be other processors that don't care about the `suppress_mailer` flag. For example, a processor that sends event information to a service like statsd.

Reply:
Yeah. You are right we should not suppress events ever. But then we need a way for stores to be able to alter already registered event handlers.